src/credentials/CHIPCert.cpp

```diff
@@ -667,61 +629,148 @@ void ValidationContext::Reset()
-
-        // Set the certificate type to CA.
-        cert.mCertType = kCertType_CA;
+        return false;
     }
 
-    // If the certificate subject contains a ChipNodeId attribute set the certificate type to Node.
-    else if (cert.mSubjectDN.mAttrOID == kOID_AttributeType_ChipNodeId)
+    if (IsChipIdX509Attr(mAttrOID))
+    {
+        return mAttrValue.mChipId == other.mAttrValue.mChipId;
+    }
+    else
     {
-        cert.mCertType = kCertType_Node;
+        return (mAttrValue.mString.mLen == other.mAttrValue.mString.mLen &&
+                memcmp(mAttrValue.mString.mValue, other.mAttrValue.mString.mValue, mAttrValue.mString.mLen) == 0);
     }
+}
+
+ChipDN::ChipDN()
+{
+    Clear();
+}
+
+ChipDN::~ChipDN()
+{
+    Clear();
+}
 
-    // If the certificate subject contains a ChipSoftwarePublisherId attribute set the certificate type to FirmwareSigning.
-    else if (cert.mSubjectDN.mAttrOID == kOID_AttributeType_ChipSoftwarePublisherId)
+void ChipDN::Clear()
+{
+    for (uint8_t i = 0; i < CHIP_CONFIG_CERT_MAX_RDN_ATTRIBUTES; i++)
     {
-        cert.mCertType = kCertType_FirmwareSigning;
+        rdn[i].Clear();
     }
+}
 
-    else
+uint8_t ChipDN::RDNCount() const
+{
+    uint8_t count;
+
+    for (count = 0; count < CHIP_CONFIG_CERT_MAX_RDN_ATTRIBUTES; count++)
     {
-        err = CHIP_ERROR_WRONG_CERT_TYPE;
+        if (rdn[count].IsEmpty())
+        {
+            break;
+        }
     }
 
+    return count;
+}
+
+CHIP_ERROR ChipDN::AddAttribute(chip::ASN1::OID oid, uint64_t chipId)
+{
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint8_t rdnCount = RDNCount();
+
+    VerifyOrExit(rdnCount < CHIP_CONFIG_CERT_MAX_RDN_ATTRIBUTES, err = CHIP_ERROR_NO_MEMORY);
+    VerifyOrExit(IsChipIdX509Attr(oid), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    rdn[rdnCount].mAttrOID           = oid;
+    rdn[rdnCount].mAttrValue.mChipId = chipId;
+
 exit:
     return err;
 }
 
-bool ChipDN::IsEqual(const ChipDN & other) const
+CHIP_ERROR ChipDN::AddAttribute(chip::ASN1::OID oid, const uint8_t * val, uint32_t valLen)
 {
-    if (mAttrOID == kOID_Unknown || mAttrOID == kOID_NotSpecified || mAttrOID != other.mAttrOID)
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint8_t rdnCount = RDNCount();
+
+    VerifyOrExit(rdnCount < CHIP_CONFIG_CERT_MAX_RDN_ATTRIBUTES, err = CHIP_ERROR_NO_MEMORY);
+    VerifyOrExit(!IsChipIdX509Attr(oid), err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(oid != kOID_NotSpecified, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    rdn[rdnCount].mAttrOID                  = oid;
+    rdn[rdnCount].mAttrValue.mString.mValue = val;
+    rdn[rdnCount].mAttrValue.mString.mLen   = valLen;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR ChipDN::GetCertType(uint8_t & certType) const
+{
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    uint8_t lCertType    = kCertType_NotSpecified;
+    bool fabricIdPresent = false;
+
+    certType = kCertType_NotSpecified;
+
+    for (uint8_t i = 0; i < RDNCount(); i++)
```

Comment:
Shouldn't RNDCount() be evaluated before the loop? Calling a for loop at every loop doesn't seem optimal in this case.
Can RDNCount() vary between each loop?